### PR TITLE
Mostly fixes for bulk-upload tests

### PR DESCRIPTION
**Pull request by pbanaszkiewicz**
Hi,

another set of fixes.

* visual fixes (like `dne` → `doesnt_exist`)
* 2 new tests

The last test, `test_upload_existing_user_existing_task`, currently fails because we apparently allow duplicate tasks.

@gvwilson and @wking: is this behavior expected? If so, I'll revert my latest commit. Otherwise we have a bug here.

### CODE REVIEW

**Comment by wking** [comment]
On Thu, Feb 26, 2015 at 01:29:16PM -0800, Piotr Banaszkiewicz wrote:
>   * test_util.py: fix code style in tests

Looks good to me.

>   * test_util.py: new test to confirm bulk-upload behavior

The test_upload_existing_user docstring doesn't match PEP 257, which
wants a single-line summary [1].

I'd also rather handle 'personal or None' and similar in Person.save.

+1 on the 400 status for invalid uploads.

>   * test_util.py: upload existing user and assign existing task

For test_upload_existing_user_existing_task, I'd expect uploading an
existing user with an existing task to be a no-op, not a failure.  In
any case, we certainly shouldn't allow duplicate tasks.  I'd recommend
using unique_together [2] to get that enforced at the database level.
For example:

  class Task(models.Model):
      '''Represent who did what at events.'''

      event = models.ForeignKey(Event)
      person = models.ForeignKey(Person)
      role = models.ForeignKey(Role)

      class Meta:
          unique_together = ('event', 'person', 'role')


[1]: https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings
[2]: https://docs.djangoproject.com/en/1.7/ref/models/options/#unique-together

**Comment by gvwilson** [comment]
Hi Piotr,
We definitely shouldn't allow duplicate tasks - the triple (event, 
person, role) should be unique (though a person might have multiple 
roles for a particular event (eg., they might be an organizer *and* an 
instructor).
Cheers,
Greg

**Comment by pbanaszkiewicz** [comment]
Hey @wking and @gvwilson,

thanks for your comments. I'll add `Meta.unique_together` soon.

> The test_upload_existing_user docstring doesn't match PEP 257, which wants a single-line summary [1].

I don't like PEP 257 for various reasons (pretty much every docstring we have violates "The docstring for a function or method should summarize its behavior and document its arguments, return value(s), side effects, exceptions raised […]"). And I don't know about any tools used for processing docstrings that do not handle this style:

```python
def func():
    """
    Description.
    """
    pass
```

("Tools for processing docstrings" is an argument used [here](https://www.python.org/dev/peps/pep-0257/#handling-docstring-indentation))

Lastly, I don't like this PEP because it does not cover a case when first sentence of the docstring is too long to be put into one line. 

I will update that docstring though.

> I'd also rather handle 'personal or None' and similar in Person.save.

I'm against that change because it doesn't fix the problem here, ie. even if we had `self.personal = self.personal or None` in `Person.save()`, we'd still have to have `personal or None` in `person_bulk_add_confirmation`. Otherwise we'd not pass verification.

**Comment by gvwilson** [comment]
Let's leave the docstrings for the post-1.0 tidy-up.

**Comment by wking** [comment]
On Thu, Feb 26, 2015 at 03:11:52PM -0800, Piotr Banaszkiewicz wrote:
> > The test_upload_existing_user docstring doesn't match PEP 257,
> > which wants a single-line summary [1].
> 
> I don't like PEP 257 for various reasons (pretty much every
> docstring we have violates "The docstring for a function or method
> should summarize its behavior and document its arguments, return
> value(s), side effects, exceptions raised […]").

It's a goal to aspire to, but "summarize in one line with an optional
blank-line-plus-further-details " is something we should be doing in
Git commits too, so I think that's pretty low-hanging fruit.  In this
case:

  Ensure existing users are assigned to new roles.

  ...more details...

fits the bill pretty well.

> Lastly, I don't like this PEP because it does not cover a case when
> first sentence of the docstring is too long to be put into one line.

If you can't summarize the function in one line, it's doing too much,
or you're trying to explain peripheral information that should be
presented in the docstring body.

> > I'd also rather handle 'personal or None' and similar in
> > Person.save.
> 
> I'm against that change because it doesn't fix the problem here,
> ie. even if we had `self.personal = self.personal or None` in
> `Person.save()`, we'd still have to have `personal or None` in
> `person_bulk_add_confirmation`. Otherwise we'd not pass
> verification.

Really?  Blank/missing URLs seem to validate fine when we're fixing
them on save (see #195).  Can you give me steps to reproduce?

**Comment by gvwilson** [comment]
Please let's worry about the features rather than the docstrings for the 
next couple of weeks.

**Comment by pbanaszkiewicz** [comment]
> > > I'd also rather handle 'personal or None' and similar in
> > > Person.save.
> >
> > I'm against that change because it doesn't fix the problem here, ie. even if we had `self.personal = self.personal or None` in `Person.save()`, we'd still have to have `personal or None` in `person_bulk_add_confirmation`. Otherwise we'd not pass verification.
>
> Really? Blank/missing URLs seem to validate fine when we're fixing them on save (see #195). Can you give me steps to reproduce?

Sure. So without `personal or None` the problem would be that we're comparing empty string with None: https://github.com/pbanaszkiewicz/amy/blob/gvwilson-cmdline-import/workshops/util.py#L124

We're not saving anything to the database, this is all happening as part of the verification step (`verify_upload_person_task`).

You can reproduce the issue by removing ` or None` in https://github.com/gvwilson/amy/pull/2/files#diff-ff40cc33cf5228ac8b825f8d4f9ef9b8R305 and then running test `BulkUploadUsersViewTestCase.test_upload_existing_user` (should fail on https://github.com/gvwilson/amy/pull/2/files#diff-f032c5050330823ed1ddb1c9298f45c5R258 due to 400 Bad request instead of 200 OK).

> Please let's worry about the features rather than the docstrings for the next couple of weeks.

Right, sorry. I'll finish up features in this PR in the morning, take care :)

**Comment by wking** [comment]
On Thu, Feb 26, 2015 at 03:31:00PM -0800, Piotr Banaszkiewicz wrote:
> > > > I'd also rather handle 'personal or None' and similar in
> > > > Person.save.
> > >
> > > I'm against that change because it doesn't fix the problem here,
> > > ie. even if we had `self.personal = self.personal or None` in
> > > `Person.save()`, we'd still have to have `personal or None` in
> > > `person_bulk_add_confirmation`. Otherwise we'd not pass
> > > verification.
> >
> > Really? Blank/missing URLs seem to validate fine when we're fixing
> > them on save (see #195). Can you give me steps to reproduce?
> 
> Sure. So without `personal or None` the problem would be that we're
> comparing empty string with None:
> https://github.com/pbanaszkiewicz/amy/blob/gvwilson-cmdline-import/workshops/util.py#L124
> 
> We're not saving anything to the database, this is all happening as
> part of the verification step (`verify_upload_person_task`).

Ah, that makes sense.  Still, 'personal' and 'family' should never be
None right?  And we should also be setting empty 'middle's to None in
Person.save (the same as we currently do for Person.email.  Otherwise
we could end up with a None middle from request.POST.getlist("middle")
with an empty-string middle from the database.

**Comment by gvwilson** [comment]
We actually have people for whom family is empty --- some south Asians 
have single names (e.g., 'Arvind'), and we record that in the 'personal' 
field.  Personally, I think life would be easier if we were each 
assigned a unique irrational number at birth to use as an identifier, 
but that proposal hasn't caught on...

**Comment by wking** [comment]
On Thu, Feb 26, 2015 at 03:49:50PM -0800, Greg Wilson wrote:
> We actually have people for whom family is empty

In that case, I'd suggest Person.family be null and blank, with the
same 'or None' handling in Person.save.

Alternatively, we could pick #66 back up, since sorting by family name
seems even less important if some folks don't have family names (and I
didn't think it was particularly important to begin with).  The
fullname/nickname fields I was recommending match up well with
Django's required get_full_name and get_short_name too.

**Comment by wking** [comment]
On Thu, Feb 26, 2015 at 03:59:20PM -0800, W. Trevor King wrote:
> Alternatively, we could pick #66 back up, …

I forgot this wasn't in the main repo.  I meant swcarpentry/amy#66.

**Comment by pbanaszkiewicz** [comment]
Hi everyone,

> And we should also be setting empty 'middle's to None in Person.save (the same as we currently do for Person.email. Otherwise we could end up with a None middle from request.POST.getlist("middle") with an empty-string middle from the database.

You're right.

> We actually have people for whom family is empty --- some south Asians have single names (e.g., 'Arvind'), and we record that in the 'personal' field.

We don't allow NULL family name, therefore Arvind would have empty string as a family name. I think this (ie. schema) should not change, because it causes some issues:

* schema rebuilding
* if-casing `Person.__str__()`

One important thing, though: we don't have any tests for this tricky case when someone doesn't have a family name. @gvwilson, could you spawn an issue for that?

Additionally, I'd like to have your confirmation that we won't be changing `Person.family` to NULL-able.

Cheers!

**Comment by pbanaszkiewicz** [comment]
Hi @wking and @gvwilson,

I added the fixes we were talking about. The only thing I'm not so sure about is whether we stick to `Person.family` being no-NULLable or do we change it to NULLable.

**Comment by wking** [comment]
On Fri, Feb 27, 2015 at 10:21:29AM -0800, Piotr Banaszkiewicz wrote:
> I added the fixes we were talking about.

fad8943 (Fix: forbid duplicate Task entries, 2015-02-27) looks good to me.

I'm not sure we need:

  self.family = self.family or ''

in 5eb4022 (Fix: empty string behavior in Person.middle and
Person.family, 2015-02-27).  How would Person.family be getting a None
value?  The ‘or None’ business works around a Django-form limitation:
the forms will only return empty strings and give you no way to
specify "unset" [1].  So we won't be getting None family values from
Django forms.  Where are they coming from?

In 74b995b2 (util.py: new verification step for duplicate Tasks,
2015-02-27), I don't think we need both ‘person = None’ and
‘person_exists’ to track missing people.  I'd also rather use
get_or_create to create Tasks, instead of having a separate check
trying to get a pre-existing Task (because I think duplicate-tasks in
the upload form should be a no-op [2]).

> The only thing I'm not so sure about is whether we stick to
> `Person.family` being no-NULLable or do we change it to NULLable.

My vote here would be to revive swcarpentry/amy#66, and have both its
name forms be non-null and non-blank.

[1]: https://code.djangoproject.com/ticket/4136
[2]: https://github.com/gvwilson/amy/pull/2#issuecomment-76291065

**Comment by pbanaszkiewicz** [comment]
> fad8943 (Fix: forbid duplicate Task entries, 2015-02-27) looks good to me. I'm not sure we need: 
> ```
> self.family = self.family or ''
> ```
> in 5eb4022 (Fix: empty string behavior in Person.middle and Person.family, 2015-02-27). How would Person.family be getting a None value? The ‘or None’ business works around a Django-form limitation: the forms will only return empty strings and give you no way to specify "unset" [1]. So we won't be getting None family values from Django forms. Where are they coming from?

You're right, now I see it doesn't make much sense. Even if someone did directly:

```
person.family = None
```

during `person.save()` there would be an error. So I'll remove that unnecessary line.

> In 74b995b (util.py: new verification step for duplicate Tasks, 2015-02-27), I don't think we need both ‘person = None’ and ‘person_exists’ to track missing people.

Yup, "person_exists" was there first, but it's no longer needed. I'll fix this.

> I'd also rather use get_or_create to create Tasks, instead of having a separate check trying to get a pre-existing Task (because I think duplicate-tasks in the upload form should be a no-op [2]).

Yeah, I was thinking about introducing a "notes" column (similarly to "errors"). This would indicate that "no operation would take place", because there's already such person and task.

But I'd rather wait for feedback from Giacomo whether or not he needs that feature.

> > The only thing I'm not so sure about is whether we stick to `Person.family` being no-NULLable or do we change it to NULLable.
>
> My vote here would be to revive swcarpentry/amy#66, and have both its name forms be non-null and non-blank.

I'd get rid of first/middle/last names, too, but:

1. Greg has some very strong points
2. This is not a place to introduce such change

**Comment by wking** [comment]
On Fri, Feb 27, 2015 at 11:34:44AM -0800, Piotr Banaszkiewicz wrote:
> > I'd also rather use get_or_create to create Tasks, instead of
> > having a separate check trying to get a pre-existing Task (because
> > I think duplicate-tasks in the upload form should be a no-op [2]).
>
> Yeah, I was thinking about introducing a "notes" column (similarly
> to "errors"). This would indicate that "no operation would take
> place", because there's already such person and task.

I don't think we need that feedback.  The user wanted to assign that
Task, and it's already in the database, so success.  Why would they
care if the task had already been in the database, as long as it's
there after their upload goes through?

> > > The only thing I'm not so sure about is whether we stick to
> > > `Person.family` being no-NULLable or do we change it to
> > > NULLable.
> >
> > My vote here would be to revive swcarpentry/amy#66, and have both
> > its name forms be non-null and non-blank.
> 
> I'd get rid of first/middle/last names, too, but:
> 
> 1. Greg has some very strong points

The last suggestions I heard on that front were:

* Adding more fields [1].
* Kicking the can down the road [2].

It's not going to take much time to land swcarpentry/amy/#66, and I'm
not a fan of *more* name fields (I think I addressed both of those
issues sufficiently in comments on that issue).  In any case, the
current name representation is causing us pain, so I think we should
fix it for good instead of patching over it here.

> 2. This is not a place to introduce such change

I'd suggest reviving and landing swcarpentry/amy/66 and then rebasing
swcarpentry/amy/#201 on top of the updated master.  Alternatively, we
could just drop those tests and leave the name handling slightly
broken, land the rest of #201, revive and land #66 on that, and then
land the fixes for name handling.

[1]: https://github.com/swcarpentry/amy/pull/66#issuecomment-67791649
[2]: https://github.com/swcarpentry/amy/pull/66#issuecomment-67992094

**Comment by pbanaszkiewicz** [comment]
> I don't think we need that feedback. The user wanted to assign that Task, and it's already in the database, so success. Why would they care if the task had already been in the database, as long as it's there after their upload goes through?

You definitely have a point here.

> I'd suggest reviving and landing swcarpentry/amy/66 and then rebasing swcarpentry/amy/#201 on top of the updated master. Alternatively, we could just drop those tests and leave the name handling slightly broken, land the rest of #201, revive and land #66 on that, and then land the fixes for name handling.

It it's current state, https://github.com/swcarpentry/amy/pull/201 plus my additions from https://github.com/gvwilson/amy/pull/2 makes bulk-upload working (even if it prevents users from adding duplicated tasks).

What I'd like to see is:

* merge this PR (https://github.com/gvwilson/amy/pull/2) - it provides a decently working web upload features on top of Greg's changes in https://github.com/swcarpentry/amy/pull/201
* merge https://github.com/swcarpentry/amy/pull/201
* probably open & merge a new PR that introduces @wking's no-op behavior (this change is probably not easy not small so I'd rather do it in separate PR)
* decide whether or not are we going to change 'first/middle/last' names into one 'name' or:
* get feedback from Giacomo regarding encoding issues and other usability issues and fix them

**Comment by wking** [comment]
On Fri, Feb 27, 2015 at 12:04:47PM -0800, Piotr Banaszkiewicz wrote:
> What I'd like to see is:
> 
> * merge this PR (https://github.com/gvwilson/amy/pull/2) - it
>   provides a decently working web upload features on top of Greg's
>   changes in https://github.com/swcarpentry/amy/pull/201
> * merge https://github.com/swcarpentry/amy/pull/201
> * probably open & merge a new PR that introduces @wking's no-op
>   behavior (this change is probably not easy not small so I'd rather
>   do it in separate PR)

Is it a big change?  I'd use get_or_create here [1] and just drop your
overly-agressive check here [2].

> * decide whether or not are we going to change 'first/middle/last'
>   names into one 'name' or:

I'm looking for Person.name and Person.nickname [3], but
swcarpentry/amy#66 was closed by the time I suggested that so I
haven't updated the code yet.

> * get feedback from Giacomo regarding encoding issues and other
>   usability issues and fix them

I think this sounds like a reasonable plan (modulo my comments above).

[1]: https://github.com/pbanaszkiewicz/amy/blob/4d121a80abfec990254d0b3101b0f861e2e7f62f/workshops/util.py#L202
[2]: https://github.com/pbanaszkiewicz/amy/blob/4d121a80abfec990254d0b3101b0f861e2e7f62f/workshops/util.py#L145
[3]: https://github.com/swcarpentry/amy/pull/66#issuecomment-67990611

File: workshops/views.py [hunk 1]
```diff
@@ -300,11 +300,12 @@ def person_bulk_add_confirmation(request):
         data_update = zip(personals, middles, families, emails, events, roles)
         for k, record in enumerate(data_update):
             personal, middle, family, email, event, role = record
+            # "field or None" converts empty strings to None values
             persons_tasks[k] = {
-                'personal': personal,
-                'middle': middle,
-                'family': family,
-                'email': email
+                'personal': personal or None,
```
Comment (gvwilson) [inline workshops/views.py]: This is the only thing I'd change - a person has to have *some* name, and I've used `personal` when they only have one.  I'll fix after merging.

**Pull request merged**